Mathlib/CategoryTheory/Category/RelCat.lean

```diff
@@ -21,75 +21,77 @@ By flipping the arguments to a relation, we construct an equivalence `opEquivale
 `RelCat` and its opposite.
 -/
 
+open Rel
+
+attribute [local simp] Rel.comp Rel.inv flip
+
 namespace CategoryTheory
 
 universe u
 
--- This file is about Lean 3 declaration "Rel".
-
 /-- A type synonym for `Type u`, which carries the category instance for which
     morphisms are binary relations. -/
 def RelCat :=
   Type u
 
-instance RelCat.inhabited : Inhabited RelCat := by unfold RelCat; infer_instance
-
-/-- The category of types with binary relations as morphisms. -/
-instance rel : LargeCategory RelCat where
-  Hom X Y := X → Y → Prop
-  id _ x y := x = y
-  comp f g x z := ∃ y, f x y ∧ g y z
-
+namespace RelCat
+variable {X Y Z : RelCat.{u}}
 
+instance inhabited : Inhabited RelCat := by unfold RelCat; infer_instance
 
-namespace RelCat
+/-- The morphisms in the relation category are relations. -/
+structure Hom (X Y : RelCat.{u}) : Type u where
+  /-- Build a morphism `X ⟶ Y` for `X Y : RelCat` from a relation between `X` and `Y`. -/
+  ofRel ::
+  /-- The underlying relation between `X` and `Y` of a morphism `X ⟶ Y` for `X Y : RelCat`. -/
+  rel : Rel X Y
 
-@[ext] theorem hom_ext {X Y : RelCat} (f g : X ⟶ Y) (h : ∀ a b, f a b ↔ g a b) : f = g :=
-  funext₂ (fun a b => propext (h a b))
+/-- The category of types with binary relations as morphisms. -/
+instance instLargeCategory : LargeCategory RelCat where
+  Hom := Hom
+  id _ := .ofRel (· = ·)
+  comp f g := .ofRel <| f.rel.comp g.rel
 
 namespace Hom
 
-protected theorem rel_id (X : RelCat) : 𝟙 X = (· = ·) := rfl
+@[ext] lemma ext (f g : X ⟶ Y) (h : f.rel = g.rel) : f = g := by
+  obtain ⟨R⟩ := f; obtain ⟨S⟩ := g; congr
+
+@[simp] protected theorem rel_id (X : RelCat.{u}) : rel (𝟙 X) = (· = ·) := rfl
```

Comment:
I'm curious if `simps` on the instance generates these, but I'm happy to leave things as is either way.